File: LeetCode/0056. Merge Intervals/Solution.java
/**
 * 170 / 170 test cases passed.
 * Runtime: 7 ms
 * Memory Usage: 46.2 MB 
 */
class Solution {
    public int[][] merge(int[][] intervals) {
        Arrays.sort(intervals, (o1, o2) -> o1[0] - o2[0]);
        int[][] ans = new int[intervals.length][2];
        int top = 0;
        for (int[] interval: intervals) {
            if (top == 0 || ans[top - 1][1] < interval[0]) {
                ans[top] = interval;
                top++;
            } else {
                ans[top - 1][1] = Math.max(ans[top - 1][1], interval[1]);
            }
        }
        return Arrays.copyOf(ans, top);
    }
}

File: LeetCode/0056. Merge Intervals/solution.cpp
/**
 * 170 / 170 test cases passed.
 * Runtime: 32 ms
 * Memory Usage: 18.5 MB 
 */
class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        sort(intervals.begin(), intervals.end());
        vector<vector<int>> ans;
        for (int i = 0; i < intervals.size(); ++ i) {
            int l = intervals[i][0], r = intervals[i][1];
            if (ans.empty() || ans.back()[1] < l) {
                ans.push_back({l, r});
            } else {
                ans.back()[1] = max(ans.back()[1], r);
            }
        }
        return ans;
    }
};
